internal/controllers/utils/hardware_utils.go

```diff
@@ -324,3 +348,147 @@ OuterLoop:
 	}
 	return nil
 }
+
+// HandleHardwareTimeout checks for provisioning or configuration timeout
+func HandleHardwareTimeout(
+	condition hwv1alpha1.ConditionType,
+	provisioningStartTime metav1.Time,
+	configurationStartTime metav1.Time,
+	timeout time.Duration,
+	currentReason string,
+	currentMessage string) (bool, string, string) {
+
+	reason := currentReason
+	message := currentMessage
+	timedOutOrFailed := false
+
+	// Handle timeout for Provisioned condition
+	if condition == hwv1alpha1.Provisioned && TimeoutExceeded(provisioningStartTime.Time, timeout) {
+		reason = string(hwv1alpha1.TimedOut)
+		message = "Hardware provisioning timed out"
+		timedOutOrFailed = true
+		return timedOutOrFailed, reason, message
+	}
+
+	// Handle timeout for Configured condition
+	if condition == hwv1alpha1.Configured && TimeoutExceeded(configurationStartTime.Time, timeout) {
+		reason = string(hwv1alpha1.TimedOut)
+		message = "Hardware configuration timed out"
+		timedOutOrFailed = true
+		return timedOutOrFailed, reason, message
+	}
+
+	// Return the current reason and message when no timeout occurred
+	return timedOutOrFailed, reason, message
+}
+
+// ValidateConfigMapFields validates the necessary fields in the hardware template ConfigMap
+func ValidateConfigMapFields(configMap *corev1.ConfigMap) error {
+	// Check for hwMgrId
+	_, exists := configMap.Data[HwTemplatePluginMgr]
+	if !exists {
+		return fmt.Errorf("missing required field '%s' in ConfigMap", HwTemplatePluginMgr)
+	}
+
+	// Check for bootInterfaceLabel
+	_, exists = configMap.Data[HwTemplateBootIfaceLabel]
+	if !exists {
+		return fmt.Errorf("missing required field '%s' in ConfigMap", HwTemplateBootIfaceLabel)
+	}
+
+	// Extract and validate node-pools-data
+	nodeGroup, err := ExtractTemplateDataFromConfigMap[[]hwv1alpha1.NodeGroup](configMap, HwTemplateNodePool)
+	if err != nil {
+		return fmt.Errorf("failed to parse 'node-pools-data' in ConfigMap %s: %w", configMap.Name, err)
+	}
+	if len(nodeGroup) == 0 {
+		return fmt.Errorf("required field 'node-pools-data' is empty in ConfigMap %s", configMap.Name)
+	}
+
+	// Validate each node group entry
+	for i, ng := range nodeGroup {
+		if ng.Name == "" {
+			return fmt.Errorf("missing 'name' in node-pools-data element at index %d", i)
+		}
+		if ng.HwProfile == "" {
+			return fmt.Errorf("missing 'hwProfile' in node-pools-data element at index %d", i)
+		}
+	}
+	return nil
+}
+
+// CompareConfigMapWithNodePool checks if there are any changes in the hardware template config map
+func CompareConfigMapWithNodePool(configMap *corev1.ConfigMap, nodePool *hwv1alpha1.NodePool, nodeGroup []hwv1alpha1.NodeGroup) (bool, error) {
+
+	// First, validate required fields in the ConfigMap
+	if err := ValidateConfigMapFields(configMap); err != nil {
```

Comment:
Can be moved to clustertemplate validation

Reply:
It has been relocated to be part of the cluster template validation.